### PR TITLE
[s3_sync] Add cache_control option

### PR DESCRIPTION
**Pull request by PapyDanone**
##### SUMMARY
Sets Cache-Control headers on uploaded objects.

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
s3_sync

##### ANSIBLE VERSION

```
ansible 2.3.1.0
  config file = 
  configured module search path = Default w/o overrides
  python version = 2.7.13 (default, Jul 18 2017, 12:33:14) [GCC 4.2.1 Compatible Apple LLVM 8.1.0 (clang-802.0.42)]
```

##### ADDITIONAL INFORMATION

* I had to add the `region` option for this to work, so I also updated the documentation.
* boto3 documentation on allowed `s3.upload_file()` arguments found [here](https://boto3.readthedocs.io/en/latest/reference/customizations/s3.html#boto3.s3.transfer.S3Transfer.ALLOWED_UPLOAD_ARGS).

cc @tedder

### CODE REVIEW

File: lib/ansible/modules/cloud/amazon/s3_sync.py [hunk 1]
```diff
@@ -105,6 +110,8 @@
     key_prefix: config_files/web
     file_change_strategy: force
     permission: public-read
+    region: us-east-1
```
Comment (tedder) [inline lib/ansible/modules/cloud/amazon/s3_sync.py]: can you pull this out? the region shouldn't be required or implied- I know you had trouble with it, but that points to a misconfigured env var or boto config file.

File: lib/ansible/modules/cloud/amazon/s3_sync.py [hunk 2]
```diff
@@ -82,6 +82,11 @@
     - For multiple patterns, comma-separate them.
     required: false
     default: ".*"
+  cache_control:
+    description:
+    - Cache-Control header set on uploaded objects.
+    - Directives are separated by commmas.
```
Comment (tedder) [inline lib/ansible/modules/cloud/amazon/s3_sync.py]: mention this is a string so it isn't implied this is a list.

File: lib/ansible/modules/cloud/amazon/s3_sync.py [hunk 3]
```diff
@@ -105,6 +110,8 @@
     key_prefix: config_files/web
     file_change_strategy: force
     permission: public-read
+    region: us-east-1
+    cache_control: public, max-age=31536000
```
Comment (tedder) [inline lib/ansible/modules/cloud/amazon/s3_sync.py]: put the value(s) in a string. personal preference but it will make the stringiness clearer.

**Review by tedder — CHANGES_REQUESTED** [review]
couple of minor changes needed.

File: lib/ansible/modules/cloud/amazon/s3_sync.py [hunk 4]
```diff
@@ -82,6 +82,11 @@
     - For multiple patterns, comma-separate them.
     required: false
     default: ".*"
+  cache_control:
+    description:
+    - Cache-Control header set on uploaded objects.
+    - Directives are separated by commmas.
+    required: false
```
Comment (pchauncey) [inline lib/ansible/modules/cloud/amazon/s3_sync.py]: here's a good place to add `version_added`

File: lib/ansible/modules/cloud/amazon/s3_sync.py [hunk 5]
```diff
@@ -25,8 +25,8 @@
 short_description: Efficiently upload multiple files to S3
 description:
      - The S3 module is great, but it is very slow for a large volume of files- even a dozen will be noticeable. In addition to speed, it handles globbing,
-       inclusions/exclusions, mime types, expiration mapping, recursion, and smart directory mapping.
-version_added: "2.3"
+       inclusions/exclusions, mime types, expiration mapping, recursion, cache control and smart directory mapping.
+version_added: "2.4"
```
Comment (pchauncey) [inline lib/ansible/modules/cloud/amazon/s3_sync.py]: This `version_added` is global for the module, not the last change to the module and should remain 2.3

Comment (tedder) [reply lib/ansible/modules/cloud/amazon/s3_sync.py]: the version_added should remain 2.3, and the cache param gets its own version_added.

Comment (tedder) [reply lib/ansible/modules/cloud/amazon/s3_sync.py]: heh, in fact, shippable caught this:

```
ERROR: lib/ansible/modules/cloud/amazon/s3_sync.py:0:0: E309 version_added for new option (cache_control) should be 2.4. Currently 0.0 (75%)
```

File: lib/ansible/modules/cloud/amazon/s3_sync.py (hunk 4 above)
Comment (PapyDanone) [reply lib/ansible/modules/cloud/amazon/s3_sync.py]: thanks @pchauncey, it makes sense

**Comment by tedder** [comment]
Great work. Looks good now. Ansibot, `shipit`.

**Comment by PapyDanone** [comment]
@tedder want me to squash all of this?

**Comment by tedder** [comment]
eh, probably wouldn't hurt.

**Review by pchauncey — APPROVED** [review]
shipit

**Review by ryansb — APPROVED** [review]

**Comment by ryansb** [comment]
For future reference: no need to squash, as we do so when merging PR's so you can skip that step. Thanks for your contribution!

**Review by ryansb — APPROVED** [review]

**Pull request merged**